wallet/recovery.go

```diff
@@ -0,0 +1,392 @@
+// have been previously found. This method ensures that the recovery state's
+// horizons properly start from the last found address of a prior recovery
+// attempt.
+func (rm *RecoveryManager) Resurrect(ns walletdb.ReadBucket,
+	scopedMgrs map[waddrmgr.KeyScope]*waddrmgr.ScopedKeyManager,
+	credits []wtxmgr.Credit) error {
+
+	// First, for each scope that we are recovering, rederive all of the
+	// addresses up to the last found address known to each branch.
+	for keyScope, scopedMgr := range scopedMgrs {
+		// Load the current account properties for this scope, using the
+		// the default account number.
+		// TODO(conner): rescan for all created accounts if we allow
+		// users to use non-default address
+		scopeState := rm.state.StateForScope(keyScope)
+		acctProperties, err := scopedMgr.AccountProperties(
+			ns, waddrmgr.DefaultAccountNum,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Fetch the external key count, which bounds the indexes we
+		// will need to rederive.
+		externalCount := acctProperties.ExternalKeyCount
+
+		// Walk through all indexes through the last external key,
+		// deriving each address and adding it to the external branch
+		// recovery state's set of addresses to look for.
+		for i := uint32(0); i < externalCount; i++ {
+			keyPath := externalKeyPath(i)
+			addr, err := scopedMgr.DeriveFromKeyPath(ns, keyPath)
+			if err != nil && err != hdkeychain.ErrInvalidChild {
+				return err
+			} else if err == hdkeychain.ErrInvalidChild {
+				scopeState.ExternalBranch.MarkInvalidChild(i)
+				continue
+			}
+
+			scopeState.ExternalBranch.AddAddr(i, addr.Address())
+		}
+
+		// Fetch the internal key count, which bounds the indexes we
+		// will need to rederive.
+		internalCount := acctProperties.InternalKeyCount
+
+		// Walk through all indexes through the last internal key,
+		// deriving each address and adding it to the internal branch
+		// recovery state's set of addresses to look for.
+		for i := uint32(0); i < internalCount; i++ {
+			keyPath := internalKeyPath(i)
+			addr, err := scopedMgr.DeriveFromKeyPath(ns, keyPath)
+			if err != nil && err != hdkeychain.ErrInvalidChild {
+				return err
+			} else if err == hdkeychain.ErrInvalidChild {
+				scopeState.InternalBranch.MarkInvalidChild(i)
+				continue
+			}
+
+			scopeState.InternalBranch.AddAddr(i, addr.Address())
+		}
+
+		// The key counts will point to the next key that can be
+		// derived, so we subtract one to point to last known key. If
+		// the key count is zero, then no addresses have been found.
+		if externalCount > 0 {
+			scopeState.ExternalBranch.ReportFound(externalCount - 1)
+		}
+		if internalCount > 0 {
+			scopeState.InternalBranch.ReportFound(internalCount - 1)
+		}
+	}
+
+	// In addition, we will re-add any outpoints that are known the wallet
+	// to our global set of watched outpoints, so that we can watch them for
+	// spends.
+	for _, credit := range credits {
+		rm.state.AddWatchedOutPoint(&credit.OutPoint)
```

Comment:
might be worth checking that this pointer won't be tied to the loop variable.

Reply:
this one should be okay, it is deferenced immediately by `AddWatchedOutPoint`

Reply:
#craefulgang